File: apps/NextJS/app/api/sosmed/posts/route.ts
import { NextResponse } from 'next/server';
import { prisma } from '@asepharyana/database';
import { getAuthenticatedUser } from '@/lib/authUtils'; // Import getAuthenticatedUser

// Initialize Prisma Client

export async function POST(request: Request) {
  try {
    // Get authenticated user
    const user = await getAuthenticatedUser();
    const userId = user?.id;

    // Parse the JSON request body
    const { content, imageUrl } = await request.json();

    // Validate request data
    if (!content || typeof content !== 'string') {
      return NextResponse.json(
        { message: 'Content is required and must be a string' },
        { status: 400 }
      );
    }

    // Check if userId is available
    if (!userId) {
      return NextResponse.json(
        { message: 'User not authenticated' },
        { status: 401 }
      );
    }

    // Create post in the database
    const newPost = await prisma.posts.create({
      data: {
        content,
        authorId: userId,
        image_url: imageUrl || '',
        userId,
      },
    });

    return NextResponse.json(
      { message: 'Post created successfully!', post: newPost },
      { status: 201 }
    );
// eslint-disable-next-line @typescript-eslint/no-unused-vars
  } catch (error) {
    // console.error('Error creating post:', error);
    return NextResponse.json(
      { message: 'Failed to create post' },
      { status: 500 }
    );
  } 
}

/**
 * Fetches all posts from the database, including associated user, comments, and likes.
 * The posts are ordered by creation date in descending order.
 * Each post's comments are enriched with user information.
 *
 * @returns {Promise<NextResponse>} A JSON response containing the sanitized posts or an error message.
 *
 * @throws {Error} If there is an error fetching the posts from the database.
 */
export async function GET() {
  try {
    // Fetch all posts from the database
    const posts = await prisma.posts.findMany({
      include: {
        user: {
          select: {
            id: true,
            name: true,
            image: true,
          },
        },
        comments: true,
        likes: true,
      },
      orderBy: {
        created_at: 'desc',
      },
    });

    // Sanitize the response
    const sanitizedPosts = await Promise.all(
      posts.map(
        async (post: {
          id: string;
          content: string;
          user: { id: string; name: string | null; image: string | null };
          comments: { id: string; content: string; userId: string }[];
          likes: { userId: string; postId: string }[];
        }) => {
          const commentsWithUser = await Promise.all(
            post.comments.map(async (comment) => {
              const user = await prisma.user.findUnique({
                where: { id: comment.userId },
                select: { id: true, name: true, image: true },
              });
              return {
                ...comment,
                user,
              };
            })
          );

          return {
            ...post,
            user: {
              id: post.user.id,
              name: post.user.name,
              image: post.user.image,
            },
            comments: commentsWithUser,
            likes: post.likes.map((like) => ({
              userId: like.userId,
              postId: like.postId,
            })),
          };
        }
      )
    );

// eslint-disable-next-line @typescript-eslint/no-unused-vars
    return NextResponse.json({ posts: sanitizedPosts }, { status: 200 });
  } catch (error) {
    // console.error('Error fetching posts:', error);
    return NextResponse.json(
      { message: 'Failed to fetch posts' },
      { status: 500 }
    );
  } 
}

export async function PUT(request: Request) {
  try {
    // Get authenticated user
    const user = await getAuthenticatedUser();
    const userId = user?.id;

    if (!user || !userId) {
      return NextResponse.json(
        { message: 'User not authenticated' },
        { status: 401 }
      );
    }
    // Parse the JSON request body
    const { id, content } = await request.json();

    // Validate request data
    if (!id || !content || typeof content !== 'string') {
      return NextResponse.json(
        { message: 'Post ID and content are required and must be valid' },
        { status: 400 }
      );
    }

    // Check if userId is available
    if (!userId) {
      return NextResponse.json(
        { message: 'User not authenticated' },
        { status: 401 }
      );
    }

    // Fetch the post to check ownership
    const post = await prisma.posts.findUnique({ where: { id } });

    if (!post || post.userId !== userId) {
      return NextResponse.json(
        { message: 'User not authorized to edit this post' },
        { status: 403 }
      );
    }

    // Update post in the database
    const updatedPost = await prisma.posts.update({
      where: { id },
      data: {
        content,
      },
    });

    return NextResponse.json(
      { message: 'Post updated successfully!', post: updatedPost },
// eslint-disable-next-line @typescript-eslint/no-unused-vars
      { status: 200 }
    );
  } catch (error) {
    // console.error('Error updating post:', error);
    return NextResponse.json(
      { message: 'Failed to update post' },
      { status: 500 }
    );
  } 
}

export async function DELETE(request: Request) {
  try {
    // Get authenticated user
    const user = await getAuthenticatedUser();
    const userId = user?.id;

    // Validate authentication
    if (!user || !userId) {
      return NextResponse.json(
        { message: 'User not authenticated' },
        { status: 401 }
      );
    }

    // Parse the JSON request body
    const { id } = await request.json();

    // Validate request data
    if (!id) {
      return NextResponse.json(
        { message: 'Post ID is required' },
        { status: 400 }
      );
    }

    // Fetch the post to check ownership
    const post = await prisma.posts.findUnique({ where: { id } });

    if (!post) {
      return NextResponse.json({ message: 'Post not found' }, { status: 404 });
    }

    if (post.userId !== userId) {
      return NextResponse.json(
        { message: 'User not authorized to delete this post' },
        { status: 403 }
      );
    }

    // Delete post from the database
    await prisma.posts.delete({ where: { id } });

    return NextResponse.json(
// eslint-disable-next-line @typescript-eslint/no-unused-vars
      { message: 'Post deleted successfully!' },
      { status: 200 }
    );
  } catch (error) {
    // console.error('Error deleting post:', error);
    return NextResponse.json(
      { message: 'Failed to delete post' },
      { status: 500 }
    );
  } 
}

File: apps/rust/src/routes/api/sosmed/posts/route.rs
use axum::{
    extract::{Query, State},
    http::{StatusCode},
    response::{IntoResponse, Response},
    Json,
};
use axum_extra::extract::cookie::CookieJar;
use serde::Deserialize;
use serde_json::json;
use std::sync::Arc;
use crate::routes::ChatState;
use crate::routes::api::user::posts_dto::{Posts, PostRequest};
use crate::routes::api::user::user_dto::User;
use crate::routes::api::user::comments_dto::Comments;
use crate::routes::api::user::likes_dto::Likes;
use crate::utils::auth::{Claims, verify_jwt};
use chrono::Utc;
use sqlx::MySqlPool;
use sqlx::FromRow;

pub async fn posts_post_handler(
    State(state): State<Arc<ChatState>>,
    cookies: CookieJar,
    Json(payload): Json<PostRequest>,
) -> Response {
    let db_pool = &state.pool;
    let jwt_secret = &state.jwt_secret;

    let token_value = match cookies.get("token") {
        Some(cookie) => cookie.value().to_string(),
        None => {
            eprintln!("No token cookie found");
            return (
                StatusCode::UNAUTHORIZED,
                Json(json!({ "message": "Authentication required" })),
            )
                .into_response();
        }
    };

    let decoded_claims = match verify_jwt(&token_value, jwt_secret).await {
        Ok(claims) => claims,
        Err(e) => {
            eprintln!("Authentication error: {:?}", e);
            return (
                StatusCode::UNAUTHORIZED,
                Json(json!({ "message": "Invalid token" })),
            )
                .into_response();
        }
    };
    let user_id = decoded_claims.user_id;

    if payload.content.is_empty() {
        return (
            StatusCode::BAD_REQUEST,
            Json(json!({ "message": "Content is required and must be a string" })),
        )
            .into_response();
    }

    let new_post_id = uuid::Uuid::new_v4().to_string();
    let created_at = Utc::now().naive_utc();
    let image_url = payload.image_url.unwrap_or_default();

    match sqlx::query_as::<_, Posts>(
        "INSERT INTO Posts (id, content, authorId, image_url, userId, created_at) VALUES (?, ?, ?, ?, ?, ?)"
    )
    .bind(&new_post_id)
    .bind(&payload.content)
    .bind(&user_id)
    .bind(&image_url)
    .bind(&user_id)
    .bind(&created_at)
    .fetch_one(db_pool.as_ref())
    .await
    {
        Ok(post) => (
            StatusCode::CREATED,
            Json(json!({ "message": "Post created successfully!", "post": post })),
        )
            .into_response(),
        Err(e) => {
            eprintln!("Error creating post: {:?}", e);
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "message": "Failed to create post" })),
            )
                .into_response()
        }
    }
}

#[derive(Debug, serde::Serialize, FromRow)]
struct PostWithUser {
    id: String,
    content: String,
    #[serde(rename = "authorId")]
    author_id: String,
    image_url: String,
    #[serde(rename = "userId")]
    user_id: String,
    created_at: chrono::NaiveDateTime,
    #[serde(rename = "user_name")]
    user_name: Option<String>,
    #[serde(rename = "user_image")]
    user_image: Option<String>,
}

#[derive(Debug, serde::Serialize, FromRow)]
struct CommentWithUser {
    id: String,
    #[serde(rename = "postId")]
    post_id: String,
    content: String,
    #[serde(rename = "userId")]
    user_id: String,
    #[serde(rename = "authorId")]
    author_id: String,
    created_at: chrono::NaiveDateTime,
    #[serde(rename = "user_name")]
    user_name: Option<String>,
    #[serde(rename = "user_image")]
    user_image: Option<String>,
}

#[derive(Debug, serde::Serialize, FromRow)]
struct LikeData {
    #[serde(rename = "userId")]
    user_id: String,
    #[serde(rename = "postId")]
    post_id: String,
}


pub async fn posts_get_handler(
    State(state): State<Arc<ChatState>>,
    cookies: CookieJar,
) -> Response {
    let db_pool = &state.pool;
    let jwt_secret = &state.jwt_secret;

    let token_value = match cookies.get("token") {
        Some(cookie) => cookie.value().to_string(),
        None => {
            eprintln!("No token cookie found");
            return (
                StatusCode::UNAUTHORIZED,
                Json(json!({ "message": "Authentication required" })),
            )
                .into_response();
        }
    };

    let decoded_claims = match verify_jwt(&token_value, jwt_secret).await {
        Ok(claims) => claims,
        Err(e) => {
            eprintln!("Authentication error: {:?}", e);
            return (
                StatusCode::UNAUTHORIZED,
                Json(json!({ "message": "Invalid token" })),
            )
                .into_response();
        }
    };
    let user_id = decoded_claims.user_id;

    match sqlx::query_as::<_, PostWithUser>(
        r#"
        SELECT
            p.id, p.content, p.authorId, p.image_url, p.userId, p.created_at,
            u.name as user_name, u.image as user_image
        FROM Posts p
        LEFT JOIN User u ON u.id = p.userId
        ORDER BY p.created_at DESC
        "#
    )
    .fetch_all(db_pool.as_ref())
    .await
    {
        Ok(posts) => {
            let mut sanitized_posts = Vec::new();
            for post in posts {
                // Fetch comments for each post
                let comments = sqlx::query_as::<_, CommentWithUser>(
                    r#"
                    SELECT
                        c.id, c.postId, c.content, c.userId, c.authorId, c.created_at,
                        u.name as user_name, u.image as user_image
                    FROM Comments c
                    LEFT JOIN User u ON u.id = c.userId
                    WHERE c.postId = ?
                    ORDER BY c.created_at DESC
                    "#
                )
                .bind(&post.id)
                .fetch_all(db_pool.as_ref())
                .await
                .unwrap_or_default();

                // Fetch likes for each post
                let likes = sqlx::query_as::<_, LikeData>(
                    "SELECT userId, postId FROM Likes WHERE postId = ?"
                )
                .bind(&post.id)
                .fetch_all(db_pool.as_ref())
                .await
                .unwrap_or_default();

                // Manually construct the desired output structure
                let user_info = json!({
                    "id": post.user_id,
                    "name": post.user_name,
                    "image": post.user_image,
                });

                let comments_with_user_info: Vec<serde_json::Value> = comments.into_iter().map(|c| {
                    json!({
                        "id": c.id,
                        "postId": c.post_id,
                        "content": c.content,
                        "created_at": c.created_at,
                        "user": {
                            "id": c.user_id,
                            "name": c.user_name,
                            "image": c.user_image,
                        }
                    })
                }).collect();

                let likes_info: Vec<serde_json::Value> = likes.into_iter().map(|l| {
                    json!({
                        "userId": l.user_id,
                        "postId": l.post_id,
                    })
                }).collect();

                sanitized_posts.push(json!({
                    "id": post.id,
                    "content": post.content,
                    "image_url": post.image_url,
                    "created_at": post.created_at,
                    "user": user_info,
                    "comments": comments_with_user_info,
                    "likes": likes_info,
                }));
            }
            (StatusCode::OK, Json(json!({ "posts": sanitized_posts }))).into_response()
        }
        Err(e) => {
            eprintln!("Error fetching posts: {:?}", e);
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "message": "Failed to fetch posts" })),
            )
                .into_response()
        }
    }
}

pub async fn posts_put_handler(
    State(state): State<Arc<ChatState>>,
    cookies: CookieJar,
    Json(payload): Json<PostRequest>,
) -> Response {
    let db_pool = &state.pool;
    let jwt_secret = &state.jwt_secret;

    let token_value = match cookies.get("token") {
        Some(cookie) => cookie.value().to_string(),
        None => {
            eprintln!("No token cookie found");
            return (
                StatusCode::UNAUTHORIZED,
                Json(json!({ "message": "Authentication required" })),
            )
                .into_response();
        }
    };

    let decoded_claims = match verify_jwt(&token_value, jwt_secret).await {
        Ok(claims) => claims,
        Err(e) => {
            eprintln!("Authentication error: {:?}", e);
            return (
                StatusCode::UNAUTHORIZED,
                Json(json!({ "message": "Invalid token" })),
            )
                .into_response();
        }
    };
    let user_id = decoded_claims.user_id;

    let Some(id) = payload.id else {
        return (
            StatusCode::BAD_REQUEST,
            Json(json!({ "message": "Post ID is required" })),
        )
            .into_response();
    };
    let content = payload.content;

    let post = match sqlx::query_as::<_, Posts>("SELECT * FROM Posts WHERE id = ?")
        .bind(&id)
        .fetch_optional(db_pool.as_ref())
        .await
    {
        Ok(Some(p)) => p,
        Ok(None) => {
            return (
                StatusCode::NOT_FOUND,
                Json(json!({ "message": "Post not found" })),
            )
                .into_response();
        }
        Err(e) => {
            eprintln!("Database error fetching post: {:?}", e);
            return (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "message": "Internal server error" })),
            )
                .into_response();
        }
    };

    if post.user_id != user_id {
        return (
            StatusCode::FORBIDDEN,
            Json(json!({ "message": "User not authorized to edit this post" })),
        )
            .into_response();
    }

    match sqlx::query_as::<_, Posts>(
        "UPDATE Posts SET content = ? WHERE id = ?"
    )
    .bind(&content)
    .bind(&id)
    .fetch_one(db_pool.as_ref())
    .await
    {
        Ok(updated_post) => (
            StatusCode::OK,
            Json(json!({ "message": "Post updated successfully!", "post": updated_post })),
        )
            .into_response(),
        Err(e) => {
            eprintln!("Error updating post: {:?}", e);
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "message": "Failed to update post" })),
            )
                .into_response()
        }
    }
}

pub async fn posts_delete_handler(
    State(state): State<Arc<ChatState>>,
    cookies: CookieJar,
    Json(payload): Json<PostRequest>,
) -> Response {
    let db_pool = &state.pool;
    let jwt_secret = &state.jwt_secret;

    let token_value = match cookies.get("token") {
        Some(cookie) => cookie.value().to_string(),
        None => {
            eprintln!("No token cookie found");
            return (
                StatusCode::UNAUTHORIZED,
                Json(json!({ "message": "Authentication required" })),
            )
                .into_response();
        }
    };

    let decoded_claims = match verify_jwt(&token_value, jwt_secret).await {
        Ok(claims) => claims,
        Err(e) => {
            eprintln!("Authentication error: {:?}", e);
            return (
                StatusCode::UNAUTHORIZED,
                Json(json!({ "message": "Invalid token" })),
            )
                .into_response();
        }
    };
    let user_id = decoded_claims.user_id;

    let Some(id) = payload.id else {
        return (
            StatusCode::BAD_REQUEST,
            Json(json!({ "message": "Post ID is required" })),
        )
            .into_response();
    };

    let post = match sqlx::query_as::<_, Posts>("SELECT * FROM Posts WHERE id = ?")
        .bind(&id)
        .fetch_optional(db_pool.as_ref())
        .await
    {
        Ok(Some(p)) => p,
        Ok(None) => {
            return (
                StatusCode::NOT_FOUND,
                Json(json!({ "message": "Post not found" })),
            )
                .into_response();
        }
        Err(e) => {
            eprintln!("Database error fetching post: {:?}", e);
            return (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "message": "Internal server error" })),
            )
                .into_response();
        }
    };

    if post.user_id != user_id {
        return (
            StatusCode::FORBIDDEN,
            Json(json!({ "message": "User not authorized to delete this post" })),
        )
            .into_response();
    }

    match sqlx::query("DELETE FROM Posts WHERE id = ?")
        .bind(&id)
        .execute(db_pool.as_ref())
        .await
    {
        Ok(_) => (
            StatusCode::OK,
            Json(json!({ "message": "Post deleted successfully!" })),
        )
            .into_response(),
        Err(e) => {
            eprintln!("Error deleting post: {:?}", e);
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "message": "Failed to delete post" })),
            )
                .into_response()
        }
    }
}
